### ISSUE TITLE
Enhancement to `unreachable` - `if-else` Control Flow

### ISSUE THREAD

**Issue by zenlyj** [PyCQA/pylint#10448]
### Current problem

The current approach towards detecting code reachability is by checking if a statement is preceded by a terminating statement (`return`, `raise`, `sys.exit()`) in the same block. Nothing wrong with that, it is straightforward and easy to maintain, and is aligned with maintainers' opinion that false negatives are preferred over false positives.

At the same time, this limits the accuracy of the check, especially for control flow constructs. An example that comes to mind is with `if-else` blocks, `unreachable` is not raised when all branches terminate.
```
def foo(flag: bool):
    if flag:
        return
    else:
        return
    print(flag)  # unreachable is not emitted
```

### Desired solution

When all `if-else` blocks terminate, any following statement should be flagged as unreachable.
```
def foo(flag: bool, flag2: bool):
    if flag:
        return
    elif flag2:
        return
    else:
        return
    print(flag)  # [unreachable]
```

When there exists at least one block that does not terminate, following statement should not be flagged as unreachable.
```
def foo(flag: bool, flag2: bool):
    if flag:
        return
    elif flag2:
        print(flag2)
    else:
        return
    print(flag)  # unreachable is not emitted
```

In my opinion, arbitrarily nested `if-else` blocks should be considered as well. I propose to use a control flow graph approach, the idea would be to generate a CFG from astroid's AST. This reduces the problem to a graph problem, where we can determine reachability by checking if there is a path between nodes.

Of course, given the dynamic nature of Python, generating a complete CFG would be extremely challenging. This is more of a first step towards introducing & generalizing basic control flow to pylint, the next steps would be to work on try-excepts and loops. The value of a CFG extends beyond the `unreachable` check, using the CFG, we can look into simplifying and improving the `VariablesChecker` as well - as discussed in #4795.

**Comment by jacobtylerwalls** [comment]
Love to see any innovation around this. @zenlyj have you spent any time with the [constraint interface](https://github.com/pylint-dev/astroid/pull/1189) in astroid? We only got as far as implementing 1 constraint, but the idea was that would help with control flow as well. If you have thoughts I'd be interested to hear. I have a "Needs astroid constraint" [label](https://github.com/pylint-dev/pylint/issues?q=is%3Aissue%20state%3Aopen%20label%3A%22Needs%20astroid%20constraint%22).

**Comment by zenlyj** [comment]
@jacobtylerwalls From my understanding, the constraint interface is designed to exclude "impossible" values during inference. I'd imagine that it would be quite expensive to use inference repeatedly across multiple checkers (where control flow matters)? That said, inference can still complement a control flow graph approach by evaluating test values in `if` conditions and pruning unreachable paths.

Honestly I am a little hesitant on implementing control flow graphs right now. Even though it is the theoretically correct way to handle control flow generically, it comes with high implementation effort, maintenance overhead (and possibly performance concerns?). Perhaps I will revisit this issue later and sketch out a potential design, but for now, I will take a stab at extending the constraint interface to cover the additional cases mentioned in https://github.com/pylint-dev/astroid/pull/1189.

**Comment by jacobtylerwalls** [comment]
That's right. At the moment I wouldn't worry about cross-checker inference, we already have a global cache for inference results. If the size of the cache needs to be adjusted or split into more caches we could optimize separately.
